skimage/transform/_seam_carving.pyx

```diff
@@ -0,0 +1,217 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: nonecheck=False
+# cython: wraparound=False
+import numpy as np
+cimport numpy as cnp
+
+
+cdef cnp.double_t DBL_MAX = np.finfo(np.double).max
+
+cdef _preprocess_image(cnp.double_t[:, :, ::1] energy_img,
+                       cnp.double_t[:, ::1] cumulative_img,
+                       cnp.int8_t[:, ::1] track_img,
+                       Py_ssize_t cols):
+    """ For each row, compute the lowest seam value for all its columns.
+
+    This function updates `cumulative_img` such that `cumulative_img[r, c]`
+    is the total energy of the lowest energy seam ending at `(r, c)`.
+
+    Parameters
+    ----------
+    energy_img : (M, N, 1) ndarray
+       The array of cost of removal of each pixel. Seam carving tries to avoid
+       pixels with high costs.
+    cumulative_img : (M, N) ndarray
+       The array to be updated with the total cost of lowest energy seams.
+    track_img : (M, N) ndarray
+        For each pixel, `track_img` stores the relative column offset in
+        the previous row which has the lowest value in `cumulative_img`. This
+        helps in in re-tracing the minimum cost seam.
+    cols : int
+        Number of columns to process.
+    """
+
+    cdef Py_ssize_t r, c, offset, c_idx
+    cdef Py_ssize_t rows = energy_img.shape[0]
+    cdef cnp.double_t min_cost = DBL_MAX
+
+    for c in range(cols):
+        cumulative_img[0, c] = energy_img[0, c, 0]
+
+    for r in range(1, rows):
+        for c in range(cols):
+            min_cost = DBL_MAX
+            for offset in range(-1, 2):
+
+                c_idx = c + offset
+                if (c_idx > cols - 1) or (c_idx < 0):
```

Comment:
Would it be worth calculating and storing `cols - 1` in the prior loop, then referencing the result?